libsolutil/Dominator.h

```diff
@@ -0,0 +1,377 @@
+			for (DfsIndex vIdx: bucket[wIdx])
+			{
+				DfsIndex uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				idom[vIdx] = (semi[uIdx] < semi[vIdx]) ? uIdx : wIdx;
+			}
+
+			// step 2
+			for (DfsIndex vIdx: predecessors[wIdx])
+			{
+				DfsIndex uIdx = eval(vIdx);
+				solAssert(uIdx <= vIdx);
+				if (semi[uIdx] < semi[wIdx])
+					semi[wIdx] = semi[uIdx];
+			}
+			bucket[semi[wIdx]].emplace_back(wIdx);
+			link(parent[wIdx], wIdx);
+			solAssert(ancestor[wIdx] == parent[wIdx]);
+		}
+
+		// step 4
+		// Compute idom in dfs order
+		idom[0] = 0;
+		for (DfsIndex wIdx: m_verticesInDFSOrder | ranges::views::drop(1) | ranges::views::transform(toDfsIndex))
+			if (idom[wIdx] != semi[wIdx])
+				idom[wIdx] = idom[idom[wIdx]];
+
+		return idom;
+	}
+
+	/// Path compression updates the ancestors of vertices along
+	/// the path to the ancestor with the minimum label value.
+	void compressPath(
+		std::vector<DfsIndex>& _ancestor,
+		std::vector<DfsIndex>& _label,
+		std::vector<DfsIndex>& _semi,
+		DfsIndex _vIdx
+	) const
+	{
+		solAssert(_ancestor[_vIdx] != std::numeric_limits<size_t>::max());
+		DfsIndex uIdx = _ancestor[_vIdx];
+		if (_ancestor[uIdx] != std::numeric_limits<size_t>::max())
+		{
+			compressPath(_ancestor, _label, _semi, uIdx);
+			if (_semi[_label[uIdx]] < _semi[_label[_vIdx]])
+				_label[_vIdx] = _label[uIdx];
+			_ancestor[_vIdx] = _ancestor[uIdx];
+		}
+		solAssert(_label[uIdx] <= _label[_vIdx]);
+	}
+
+	/// Build dominator tree from the immediate dominators set.
+	/// The function groups all the vertices ids that are immediately dominated by a vertex.
+	void buildDominatorTree()
+	{
+		// m_immediateDominators is guaranteed to have at least one element after findingDominators() is executed.
+		solAssert(m_immediateDominators.size() > 0);
+		solAssert(m_immediateDominators.size() == m_verticesInDFSOrder.size());
+		solAssert(m_immediateDominators[0] == 0);
+
+		// Ignoring the entry node since no one dominates it.
+		for (DfsIndex index = 1; index < m_verticesInDFSOrder.size(); ++index)
+		{
+			VId vId = m_verticesInDFSOrder[index];
+			solAssert(m_dfsIndexByVertexId.count(vId));
+			solAssert(index == m_dfsIndexByVertexId.at(vId));
+			VId idomId = m_verticesInDFSOrder[m_immediateDominators[index]];
+			solAssert(m_immediateDominators[index] < index);
+			m_dominatorTree[idomId].emplace_back(vId);
+		}
+	}
+
+	/// Keep the list of vertices ids in the DFS order.
+	/// The entry vertex is the first element of the vector.
+	/// The index of the other vertices is based on the order they are visited during the DFS.
+	/// i.e. m_verticesInDFSOrder[i] is the vertex ID whose DFS index is i.
+	///
+	/// DFS index -> Vertex id
+	std::vector<VId> m_verticesInDFSOrder;
+
+	/// Maps a Vertex id to its DFS order index.
+	/// NOTE: The CFG can be disconnected, so the vertex Ids are not necessarily contiguous.
+	///
+	/// Vertex id -> DFS index
+	std::map<VId, DfsIndex> m_dfsIndexByVertexId;
```

Comment:
The naming here is a bit inconsistent, with it being sometimes `DFS` and sometimes `Dfs`. I'd generally prefer `DFS` (and also `ID`), because this is the way we seem to be doing it most of the time, but I'd really take any of them over a random mix :)

Reply:
Also, the files should be renamed to match the new name of the class (`Dominator.h` -> `DominatorFinder.h`, `DominatorTest.cpp` -> `DominatorFinderTest.cpp`).

Reply:
Ok. I agree, but do you also mean to not use Camel Case for `ID` and `DFS` on variable names of only on the comments? I can do the change either way.

Reply:
I meant names. Comments I fixed already since that was a no-brainer.

And yeah, I can live with it either way, but it's just one of those things that we keep flipping between randomly all the time and is really annoying to keep track of. Especially when you need to modify the code later and can't decide which version to stick to. Similar kind of annoyance as the eternal `optimize`/`optimise` thing :)

In any case this is also something that is probably best off addressed separately. For now we should just merge the PR :)